design/service-apis-implementation.md

```diff
@@ -0,0 +1,342 @@
+Currently, Contour provides "TCP Proxying" that allows the forwarding of TLS streams based on the SNI information, which is precisely what the Service APIs TLSRoute object is for.
+If Project Contour (the organisation) does add support for TCP and UDP forwarding, it will not be in the `projectcontour/contour` repo, but will be a separate repo.
+
+This design is intended to cover the initial `v1alpha1` release of the Service APIs. We will aim to implement the core featureset of the APIs at that point.
+We will then work with the upstream community on features that Contour and HTTPProxy currently support, but the Service APIs do not, and how best to represent those features in the APIs.
+So, there are some features that are currently gaps for Contour (wildcard domain names, and exact path matching seem like big ones),
+and some that Contour supports that the Service APIs do not (websockets, configurable timeouts, header replacement, external auth, rate limiting, and so on).
+
+Our eventual ideal is that HTTPProxy and Service-APIs will have feature parity.
+Practically, there may be times when HTTPProxy can move faster than Service-APIs because we have a smaller feature surface.
+We see this as a chance for HTTPProxy to test out ideas for Service-APIs functionality, and provide feedback on what features should be in the core, extension, and implementation-specific parts of the API.
+There should never be annotations required for functionality on Service-APIs objects.
+
+
+## Goals
+- Define a data model and implementation for Contour's service-apis support, covering the v1alpha1 version of the service-apis.
+- Layer 7 support only, which means HTTPRoutes, TLSRoutes only.
+
+## Non Goals
+- No TCPRoute or UDPRoute support, that is, no support for arbitrary TCP/UDP forwarding.
+
+
+## High-Level Design
+
+### Service APIs implementation surface area
+
+Contour's support for the Service APIs will include support for the HTTPRoute and TLSRoute objects only, not any other type of Route - this includes no TCPRoute or UDPRoute support.
+Contour is a layer 7 ingress controller, and the layer 4 load balancing implies by TCPRoute and UDPRoute is out of scope for this tool (that is, `projectcontour/contour`).
+Project Contour (the organisation) may investigate a Service-APIs based layer 4 solution in the future, but that effort will not be in `projectcontour/contour`.
+
+### Contour Gateway model
+
+Contour considers a Gateway to describe a single Envoy deployment, deployed and managed by something outside of itself.
+Contour expects to be supplied with the full name (name and namespace) of a single Gateway, which it will watch and update the status of.
+This will be a configuration file parameter.
+When a Gateway is not supplied, Contour will not do Service-APIs processing, or in other words, the existing `--experimental-service-apis` flag will be obsoleted and removed).
```

Comment:
+1 on removing the flag. 

Should we default the name/namespace of a gateway (i.e. contour/projectcontour) so given a generic deployment, someone could use that as a default and not require a config file change?

Contour can still check if the service-apis CRDs exist in the cluster and use that as a key for setting up informers.  I just merged https://github.com/projectcontour/contour/pull/3283 which acts this way, but is different than this design doc, but @skriss asked which is a good point.

Reply:
So, the flow if you want to use the service-apis is this:
- install the objects in the cluster
- create a Gateway for Contour to reference
- Configure Contour to use that

If you don't do all those three things, then no service-apis.

That does make sense, although it means that if you install the objects and don't change the default, we should end up logging something saying "hey, you asked me to look for a Gateway called `projectcontour/contour`, but it doesn't exist". Not a huge deal.

Reply:
Right now I have a default programmed, but we could remove that, just thought it might be nice to have some way of allowing someone to create a Gateway without needing to reconfigure / restart Contour. 

This came up in a review with @skriss and thought it was a good point that what I PR'd conflicted with this design and thought we should clear up. =)